flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/NettyShuffleUtils.java

```diff
@@ -0,0 +1,138 @@
+            final int sortShuffleMinBuffers,
+            final int numSubpartitions,
+            final ResultPartitionType type) {
+        int min =
+                type.isBlocking() && numSubpartitions >= sortShuffleMinParallelism
+                        ? sortShuffleMinBuffers
+                        : numSubpartitions + 1;
+        int max =
+                type.isBounded()
+                        ? numSubpartitions * numBuffersPerChannel + numFloatingBuffersPerGate
+                        : Integer.MAX_VALUE;
+        return Pair.of(min, max);
+    }
+
+    public static int computeNetworkBuffersForAnnouncing(
+            final int numBuffersPerChannel,
+            final int numFloatingBuffersPerGate,
+            final int sortShuffleMinParallelism,
+            final int sortShuffleMinBuffers,
+            final int numTotalInputChannels,
+            final int numTotalInputGates,
+            final Map<IntermediateDataSetID, Integer> subpartitionNums,
+            final Map<IntermediateDataSetID, ResultPartitionType> partitionTypes) {
+
+        // Each input channel will retain N exclusive network buffers, N = numBuffersPerChannel.
+        // Each input gate is guaranteed to have a number of floating buffers.
+        int requirmentForInputs =
+                numBuffersPerChannel * numTotalInputChannels
+                        + numFloatingBuffersPerGate * numTotalInputGates;
+
+        int requirementForOutputs = 0;
+        for (IntermediateDataSetID dataSetId : subpartitionNums.keySet()) {
+            int numSubs = subpartitionNums.get(dataSetId);
+            checkArgument(partitionTypes.containsKey(dataSetId));
+            ResultPartitionType partitionType = partitionTypes.get(dataSetId);
+
+            requirementForOutputs +=
+                    getNumBuffersToAnnounceForResultPartition(
+                            partitionType,
+                            numBuffersPerChannel,
+                            numFloatingBuffersPerGate,
+                            sortShuffleMinParallelism,
+                            sortShuffleMinBuffers,
+                            numSubs);
+        }
+
+        return requirmentForInputs + requirementForOutputs;
+    }
+
+    private static int getNumBuffersToAnnounceForResultPartition(
+            ResultPartitionType type,
+            int numBuffersPerChannel,
+            int floatingBuffersPerGate,
+            int sortShuffleMinParallelism,
+            int sortShuffleMinBuffers,
+            int numSubpartitions) {
+
+        Pair<Integer, Integer> minAndMax =
+                getMinMaxNetworkBuffersPerResultPartition(
+                        numBuffersPerChannel,
+                        floatingBuffersPerGate,
+                        sortShuffleMinParallelism,
+                        sortShuffleMinBuffers,
+                        numSubpartitions,
+                        type);
+
+        final int ret;
+        if (type.isPipelined()) {
```

Comment:
I think we first need to document that our goal is to avoid buffer request timeout (see FLINK-12852).
For pipelined shuffle type, the floating buffers may not be returned in time due to back pressure so we need to include all the floating buffers in the announcement, i.e. we should take the max value.
For blocking shuffle type, it is back pressure free and and floating buffers can be recycled in time, so that the minimum required buffers would be enough.

Reply:
Updated.